src/backends/simple.ts

```diff
@@ -0,0 +1,442 @@
+        }
+        else {
+            throw new Error(`Unexpected interpolation mode: ${interpolateMode}`);
+        }
+    }
+
+    filterSpec = [...filterSpec];
+
+    const filterSpecStr = filterSpec.join(', ') + '[out0:v]';
+
+    LOG && console.log(`filterSpec: ${filterSpecStr}`);
+
+    return beamcoder.filterer({
+        filterType: 'video',
+        inputParams: [
+            {
+                name: 'in0:v',
+                width: stream.codecpar.width,
+                height: stream.codecpar.height,
+                pixelFormat: stream.codecpar.format,
+                timeBase: stream.time_base,
+                pixelAspect: stream.sample_aspect_ratio,
+            },
+        ],
+        outputParams: [
+            {
+                name: 'out0:v',
+                pixelFormat: outputPixelFormat,
+            },
+        ],
+        filterSpec: filterSpecStr,
+    });
+};
+
+const MAX_PACKET_READS = 1000;
+const STREAM_TYPE_VIDEO = 'video';
+
+/**
+ * A simple extractor that uses beamcoder to extract frames from a video file.
+ */
+export class SimpleExtractor extends BaseExtractor implements Extractor {
+    /**
+     * The demuxer reads the file and outputs packet streams
+     */
+    demuxer: Demuxer = null;
+
+    /**
+     * The decoder reads packets and can output raw frame data
+     */
+    decoder: Decoder = null;
+
+    /**
+     * Packets can be filtered to change colorspace, fps and add various effects. If there are no color space changes or
+     * filters, filter might not be necessary.
+     */
+    filterer: Filterer = null;
+
+    /**
+     * This is where we store our filtered frames closest to the target time. We hang on to them for two reasons:
+     * 1. so we can return them if we get a request for the same time again
+     * 2. so we can return frames close the end of the stream. When such a frame is requested we have to flush (destroy)
+     * the encoder to get the last few frames. This avoids having to re-create an encoder.
+     */
+    filteredFrames: undefined[] | DecodedFrames = [];
+
+    /**
+     * This contains the last raw frames we read from the demuxer. We use it as a starting point for each new query. We
+     * do this ensure we don't skip any frames.
+     */
+    frames = [];
+
+    /**
+     * This contains the last packet we read from the demuxer. We use it as a starting point for each new query. We do
+     * this ensure we don't skip any frames.
+     */
+    packet: Packet = null;
+
+    /**
+     * The last target presentation timestamp (PTS) we requested. If we never requested a time(stamp) then this
+     * value is null
+     */
+    previousTargetPTS: null | number = null;
+
+    /**
+     * The number of threads to use for decoding
+     */
+    threadCount = 8;
+
+    /**
+     * The index of the video stream in the demuxer
+     * @param args
+     */
+    streamIndex = 0;
+
+    static async create(args: ExtractorArgs): Promise<Extractor> {
```

Comment:
```suggestion
    /**
     * Encoder/Decoder constuction is async, so it can't be put in a regular constructor.
     *
     * Use and await this method to generate an extactor.
     */
    static async create(args: ExtractorArgs): Promise<Extractor> {
```

Reply:
To answer @stepancar's question here: https://github.com/Lumen5/framefusion/pull/27/files#r1184435510